src/ccontrol/MergingHandler.cc

```diff
@@ -38,22 +43,269 @@
 #include "rproc/InfileMerger.h"
 #include "util/Bug.h"
 #include "util/common.h"
 #include "util/StringHash.h"
 
 using lsst::qserv::proto::ProtoHeader;
+using lsst::qserv::proto::ProtoHeaderWrap;
 using lsst::qserv::proto::ProtoImporter;
 using lsst::qserv::proto::Result;
 using lsst::qserv::proto::WorkerResponse;
+using lsst::qserv::replica::HttpClient;
 
 using namespace std;
 
 namespace {
 LOG_LOGGER _log = LOG_GET("lsst.qserv.ccontrol.MergingHandler");
+
+string xrootdStatus2str(XrdCl::XRootDStatus const& s) {
+    return "status=" + to_string(s.status) + ", code=" + to_string(s.code) + ", errNo=" + to_string(s.errNo) +
+           ", message='" + s.GetErrorMessage() + "'";
+}
+
+/**
+ * Extract the file path (including both slashes) from the XROOTD-style URL.
+ * Input:
+ *   @code
+ *   "xroot://<server>:<port>//<path>""
+ *   @code
+ * Output:
+ *   @code
+ *   "//<path>""
+ *   @code
+ */
+string xrootUrl2path(string const& xrootUrl) {
+    string const delim = "//";
+    auto firstPos = xrootUrl.find(delim, 0);
+    if (string::npos != firstPos) {
+        // Resume serching at the first character following the delimiter.
+        auto secondPos = xrootUrl.find(delim, firstPos + 2);
+        if (string::npos != secondPos) {
+            return xrootUrl.substr(secondPos);
+        }
+    }
+    throw runtime_error("MergingHandler::" + string(__func__) + " illegal file resource url: " + xrootUrl);
+}
+
+bool readXrootFileResourceAndMerge(lsst::qserv::proto::Result const& result,
+                                   function<bool(char const*, uint32_t)> const& messageIsReady) {
+    string const context = "MergingHandler::" + string(__func__) + " ";
+
+    // Extract data from the input result object before modifying the one.
+    string const xrootUrl = result.fileresource_xroot();
+
+    // The algorithm will read the input file to locate result objects containing rows
+    // and call the provided callback for each such row.
+    XrdCl::File file;
+    XrdCl::XRootDStatus status;
+    status = file.Open(xrootUrl, XrdCl::OpenFlags::Read);
+    if (!status.IsOK()) {
+        LOGS(_log, LOG_LVL_ERROR,
+             context << "failed to open " << xrootUrl << ", " << xrootdStatus2str(status));
+        return false;
+    }
+
+    // Temporary buffer for messages read from the file. The buffer will be (re-)allocated
+    // as needed to get the largest message. Note that a size of the messages won't exceed
+    // a limit set in ProtoHeaderWrap::PROTOBUFFER_HARD_LIMIT.
+    unique_ptr<char[]> buf;
+    size_t bufSize = 0;
+
+    uint64_t offset = 0;  // A location of the next byte to be read from the input file.
+    bool success = true;
+    try {
+        while (true) {
+            // Read the frame header that carries a size of the subsequent message.
+            uint32_t msgSizeBytes = 0;
+            uint32_t bytesRead = 0;
+            status = file.Read(offset, sizeof(uint32_t), reinterpret_cast<char*>(&msgSizeBytes), bytesRead);
```

Comment:
If the size is not correct then the algorithm would fail to parse the subsequent Protobuf message. So, I think, extra reinforcement is needed here. I'm not seeing such reinforcement in the existing SSI-streaming-based result delivery protocol.